src/odemis/driver/pigcs.py

```diff
@@ -1971,6 +1967,24 @@ def __init__(self, busacc, address=None, axes=None, auto_suspend=10, suspend_mod
             self._encoder_mng[a] = t
             t.start()
 
+            # TODO: Check if there was any need to send the RON command via SetReferenceMode in the startServo method.
+            #  @Eric was there any reason for this? It seems to work fine in the tests without.
+            # Referencing:
+            # When sending the 'RON' command with False (RON disabled), the axes are not actually referenced.
+            # When setting a position (SetPosition()) with RON disabled ("RON", False), the axis will report
+            # to be referenced, though it is actually not. This allows to move without referencing beforehand.
+            # To allow (absolute) moves, even if it's not actually referenced: pass False
+            # To do proper referencing, check if has reference limits and/or switch: pass True
+            # TODO Eric: Is there a situation where we explicitly want to pass False here even
+            #  if we have a reference switch/limits?
+            #  Is there any controller operated in closed-loop but without ref sensor?
+            self.SetReferenceMode(a, self._hasLimitSwitches[a] or self._hasRefSwitch[a])
+            if self._hasLimitSwitches[a] or self._hasRefSwitch[a]:
+                # TODO problem is that we now call it without a lock
+                #  -> we dont't wait until finished and bus gets reported that it it not referenced
+                #  cannot call doReference here as method on Bus
+                self.startReferencing(a)
+
```

Comment:
This doesn't work properly as we do not wait until referencing finished. Any suggestions how to implement or call at a different place?

Reply:
Anyway, as a basic rule, we never move the hardware without the user explicitly requesting it. So normally we never do any referencing at driver initialisation. (In the case of this microscope, if it's actually always safe to do, we can still do it early via the GUI or a new special init_on_ref argument).

I'd suggest the following. The driver distinguishes between two kinds of hardware: the one which can reference (ie, has a Limit or Ref switch), and the one which cannot. For the later case, the code should behave exactly as before (eg, RON False all the time, and reports it cannot be referenced), and therefore we should be confident not breaking behaviour on existing hardware (see note below though). In the case the hardware is referenceable:
* at init, and until it's been successfully referenced, it should behave like non-referenceable hardware. When isReferenced() is called, it should return False (so you might need to *not* call FRF?, but just store the status internally)
* After being referenced, it should use "RON True" (or maybe not needed, as referencing already does it?) and isReferenced() should report it's been referenced.
Maybe at init it's possible to check whether the hardware has already been referenced or not, using a command like "RON?" . In the worse case, I'd suggest to assume at init that it's not referenced, and hopefully in reality, the backe-end is sufficiently rarely restarted that's it'd become an issue.

Note: some of the current hardware might report the presence of reference or limit switches *even if* they don't support referencing (because physically they have the switch, it just doesn't work). We need to check. If so, then we'll need to bring some extra cleverness...